Feat: add program

SoftAvocado opened the pull request:
Added library and tests.

SoftAvocado committed:
feat: add program

SoftAvocado committed:
Create dotnet.yml

SoftAvocado committed:
Update dotnet.yml

SoftAvocado committed:
Update dotnet.yml

SoftAvocado committed:
Update dotnet.yml

SoftAvocado committed:
Fixes

SoftAvocado committed:
Update dotnet.yml

SoftAvocado committed:
Fixes

Koval1997 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+
+namespace BowlingCalculator
+{
+    public class Calculator
```
В ООП следует создавать отдельный файл под каждый класс. Т.е. файл Game.cs в нем класс и т.д.

Koval1997 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+    {
+        Game game;
+
+        public List<int> calculate_Throws(List<int> throws)
```
Имена методов класса пишутся с большой буквы и без нижних подчеркиваний calculate_Throws -> CalculateThrows

vvvictor07 replied:
PascalCase

Koval1997 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+
+        public List<int> calculate_Throws(List<int> throws)
+        {
+            this.game = new Game();
```
this, по-моему, уже можно не писать

vvvictor07 replied:
лучше не надо

Koval1997 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+        public List<int> calculate_Throws(List<int> throws)
+        {
+            this.game = new Game();
+            List<int> calculator_results = new List<int>();
```
локальные переменные тоже именуются в camelCase,  calculator_results -> calculatorResults

vvvictor07 replied:
https://medium.com/@alivander/camel-pascal-snake-case-%D0%B8-%D0%B4%D1%80%D1%83%D0%B3%D0%B8%D0%B5-%D1%81%D1%82%D0%B8%D0%BB%D0%B8-%D0%BD%D0%B0%D0%BF%D0%B8%D1%81%D0%B0%D0%BD%D0%B8%D1%8F-288ec62ca0d0

Koval1997 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+    class Game
+    {
+        private List<Frame> results = new List<Frame>();
+        static int frame_id_counter = 1;
```
Static, наверное, здесь не стоит использовать. Т.к. static определяет член, который принадлежит не к объекту типа Game, а к самому типу

Koval1997 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+                }
+            }
+
+            public bool Get_isFrameEnded()
```
Те методы, которые возвращают bool, т.е. проверяют какое-то условие и т.д. принято называть с is, без get как в данном случае, т.е. просто isFrameEnded. А т.к. этот метод и другие только возвращают значения свойств, то можно это все сделать через get/set

vvvictor07 replied:
Example:
```c#
public bool IsSpare { get; private set; }
```

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+﻿using System;
```
Unused import

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace BowlingCalculator
```
It is better if namespace name is equal to folder name
```suggestion
namespace Bowling_Calculator
```

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+        {
+            this.game = new Game();
+            List<int> calculator_results = new List<int>();
+            for (int i = 0; i < throws.Count; i++)
```
It isn't necessary but I think that better use `foreach` cycle here

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+            return calculator_results;
+        }
+
+        public List<int> Get_FrameResults(int frame_id)
```
Fix according to [the naming contentions](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions), please

vvvictor07 commented on .github/workflows/dotnet.yml:
```diff
@@ -0,0 +1,25 @@
+name: .NET
```
I need the `Readme.md` file in the root folder. This file should contain description of the task and other information if needed

SoftAvocado replied:
There is a readme file in the root folder named README.md with a short description of the task.

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+
+        class Frame
+        {
+            int id;
```
This field could be removed

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+                this.isStrike = false;
+                this.isSpare = false;
+                this.isBonusThrow = false;
+                this.isFrameEnded = false;
```
Bool is initialized by default value (false), so this is redundant.

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+                this.isBonusThrow = false;
+                this.isFrameEnded = false;
+                this.frameScore = 0;
+                this.total = 0;
```
The same as with bool

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+        class Frame
+        {
+            int id;
+            bool isStrike;
+            bool isSpare;
+            bool isBonusThrow;
+            bool isFrameEnded;
+            List<int> resultThrows = new List<int>();
+            List<int> addedThrows = new List<int>();
+            int frameScore;
+            int total;
```
All this fields is private so them names should start with `_` (underscore) by naming convention. Fix please.

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+                this.resultThrows.Add(resultThrow);
+                this.Add_FrameScore(resultThrow);
+                this.total += resultThrow;
+                if (this.isFrameEnded == false)
```
This could be converted to code below to reduce nesting
```c#
                if (isFrameEnded != false)
                {
                    return;
                }
```

vvvictor07 commented on Bowling_Calculator/BowlingCalculator.cs:
```diff
@@ -0,0 +1,283 @@
+                        this.isFrameEnded = true;
+                    }
+                    // spare bonus end
+                    if ((this.isBonusThrow) && (this.isSpare) && (this.addedThrows.Count == 3))
+                    {
+                        this.isFrameEnded = true;
+                    }
+                    // strike bonus end
+                    else if ((this.isBonusThrow) && (this.isStrike) && (this.addedThrows.Count == 3))
+                    {
+                        this.isFrameEnded = true;
+                    }
+                    else if (this.isBonusThrow)
+                    {
+                        this.isFrameEnded = false;
+                    }
```
This could be converted to switch-case statement (yes it isn't python, we have switch-case :-) here) 
But it still strange, may we could reduce this chain?

SoftAvocado replied:
I don't think switch-case will be usefull here as all three conditions do not follow the same pattern. Though the chain can be reduced by combining first two expressions, thank you.

SoftAvocado committed:
Delete BowlingCalculator.cs

Deleted file with all classes to upload classes in seperate files.

SoftAvocado committed:
Feat: modified program

Fixes:
- all classes have been seperated into different files
- all variables, methodes and classes names are fixed according naming contentions, PascalCase was used
- namespace was renamed as folder name
- all unused imports have been removed
- getters and setter have been modified
- unsed field 'id' was removed from Frame class
- added FrameIDCounter reload at the end of the game

SoftAvocado committed:
docs: update readme

Added game description and discription of current algorithm.

SoftAvocado committed:
feat: modifed program

Program now works without static variable frameIdCounter.

SoftAvocado committed:
docs: update readme

Koval1997 commented on Bowling_Calculator/Game.cs:
```diff
@@ -0,0 +1,136 @@
+
+        void UpdateTotalScoreOfFrame(int frameId)
+        {
+            if (frameId + 1 == 1)
```
Кажется, что проще написать if (frameId == 0) {}

Koval1997 commented on Bowling_Calculator/Game.cs:
```diff
@@ -0,0 +1,136 @@
+        int DefineCurrentFrame()
+        {
+            int frameId;
+            if ((_results.Count == 0)||(_results[(_results.Count) - 1].IsFrameEnded))
```
Вот эту штуку _results[(_results.Count) - 1] лушче вынести в переменную выше и обозвать так, чтобы было понятно :)

SoftAvocado replied:
@Koval1997, можно вынести это значение в переменную, но для объявления этой переменной должно выполняться условие _results.Count != 0, из-за чего кода становится больше и, по-моему, он становится менее понятным. Лучше ли стало? Может, просто добавить комментарий с описанием условия? 
```
        int DefineCurrentFrame()
        {
            int frameId = _results.Count - 1;
            if (_results.Count == 0)
            {
                frameId = _results.Count;
                _results.Add(new Frame());
            }
            else
            {
                Frame lastFrame = _results[(_results.Count) - 1];
                if (lastFrame.IsFrameEnded)
                {
                    frameId = _results.Count;
                    _results.Add(new Frame());
                }
            }
            return frameId;
        }
```

Koval1997 commented on BowlingCalculator_UnitTests/BowlingCalculatorTests.cs:
```diff
@@ -0,0 +1,106 @@
+        [Test]
+        public void calculate_Throws_standartThrows_lastThrowIsSpare()
+        {
+            List<int> throws = new List<int>() { 10, 7, 3, 7, 2, 9, 1, 10, 10, 10, 2, 3, 6, 4, 7, 3, 3 };
```
Еще в прошлый раз забыли отметить, что для локальных переменных можно не указывать типы при объявлении, а использовать var, при компиляции тип определится сам

Koval1997 commented on BowlingCalculator_UnitTests/BowlingCalculatorTests.cs:
```diff
@@ -0,0 +1,106 @@
+        {
+            List<int> throws = new List<int>() { 10, 7, 3, 7, 2, 9, 1, 10, 10, 10, 2, 3, 6, 4, 7, 3, 3 };
+            List<int> expexted_result = new List<int>() { 10, 24, 30, 44, 46, 55, 56, 76, 96, 126, 132, 138, 144, 148, 162, 165, 168 };
+            List<int> actual_result = new List<int>();
```
Эту переменную можно не объявлять заранее, а сразу там, где присвоить в нее результат, в строке 17

SoftAvocado committed:
feat: modified program

Fixes:
- added comment to DefineCurrentFrame function
- simplified unit tests' code

Commit:
docs: add CI/CD information
